### PR TITLE
support calibration dataset concat

### PR DESCRIPTION
**Pull request by yintong-lu**
dataset = datasetname1:split=train+val:num=256,datasetname2:concat=True:num=128,datasetname3:concat

We use ',' to split datasets and ':' to split parameters of a dataset.

### CODE REVIEW

File: auto_round/calib_dataset.py [hunk 1]
```diff
@@ -251,11 +253,38 @@ def filter_func(example):
...
     datasets = []
     for name in dataset_names:
         split = None
+        do_concat = False
         if ":" in name:
-            name, split = name.split(":")
+            # name, split = name.split(":")
+            split_list = name.split(":")
+            name, split_list = name.split(":")[0], name.split(":")[1:]
+            do_concat = 'concat' in split_list
```
Comment (wenhuach21) [inline auto_round/calib_dataset.py:285]: bug, move do_concat into if

**Comment by wenhuach21** [comment]
please align the readme in example

**Comment by wenhuach21** [comment]
@mxjmtxrm is this one meet your requirements?

**Comment by mxjmtxrm** [comment]
It seems wrong.
The attention mask should be like [1,1,1,1,2,2,2,2,3,3,0,0,0] as 1,2,3 are different sequences(0 is padding token).

**Comment by wenhuach21** [comment]
> It seems wrong. The attention mask should be like [1,1,1,1,2,2,2,2,3,3,0,0,0] as 1,2,3 are different sequences(0 is padding token). Ref:https://github.com/MeetKai/functionary/tree/main/functionary/train/packing#packing-inputs-without-cross-contamination-attention
Thank you for the information; this is my first time learning about this method. However, we will not going to support it because it is not the standard practice and has some limitations that require aligning and patching some functions in the model fine .

**Comment by mxjmtxrm** [comment]
Have you tried dataset padding? like input_ids: [66,66,66,0,0] attention mask: [1,1,1,0,0].
I met index out of range error in `sampling_inputs` function.

**Comment by wenhuach21** [comment]
> Have you tried dataset padding? like input_ids: [66,66,66,0,0] attention mask: [1,1,1,0,0]. I met index out of range error in `sampling_inputs` function.

Yes, https://github.com/intel/auto-round/blob/main/auto_round/calib_dataset.py#L307 a simple workaround is just set the all last attention mask to 0

The root cause is that we need to cache the intermediate attention mask of the first block for each sample. However, if all the values of input attention mask for samples in the batch are 1, then some models will clear the intermediate attention mask, setting it to None, which causes us to drop these samples as we could not batch them with samples with not-none attention mask

**Comment by wenhuach21** [comment]
attention_mask = attention_mask[:seqlen]
attention_mask[-1]=0

**Comment by mxjmtxrm** [comment]
Works for me. Thanks.

**Comment by wenhuach21** [comment]
is this pr ready to merge? @yintong-lu

**Comment by yintong-lu** [comment]
> is this pr ready to merge? @yintong-lu

Yes, please review it and let me know if there's any thing I might change.

File: examples/language-modeling/README.md [hunk 2]
```diff
@@ -36,6 +36,7 @@ def customized_data_with_tokenizer(tokenizer, seqlen=2048):
 ~~~
 
 Combination of different datasets has been supported, --dataset "./tmp.json,NeelNanda/pile-10k:train, mbpp:train+validation+test". Please note that samples with sequence length < args.seqlen will be dropped.
+We also support parametrization of calibration datasets by using --dataset "datasetname1:split=train+val:num=256,datasetname2:concat=True:num=128,datasetname3:concat". Please use ',' to split datasets and ':' to split parameters of a dataset.
```
Comment (wenhuach21) [inline examples/language-modeling/README.md]: consolidate the  following readme


Combination of different datasets has been supported, --dataset "./tmp.json,NeelNanda/pile-10k:train, mbpp:train+validation+test". Please note that samples with sequence length < args.seqlen will be dropped. We also support parametrization of calibration datasets by using --dataset "datasetname1:split=train+val:num=256,datasetname2:concat=True:num=128,datasetname3:concat". Please use ',' to split datasets and ':' to split parameters of a dataset.

**Review by wenhuach21 — APPROVED** [review]

File: examples/language-modeling/README.md [hunk 3]
```diff
@@ -35,8 +35,8 @@ def customized_data_with_tokenizer(tokenizer, seqlen=2048):
     return tokens
 ~~~
 
-Combination of different datasets has been supported, --dataset "./tmp.json,NeelNanda/pile-10k:train, mbpp:train+validation+test". Please note that samples with sequence length < args.seqlen will be dropped.
-We also support parametrization of calibration datasets by using --dataset "datasetname1:split=train+val:num=256,datasetname2:concat=True:num=128,datasetname3:concat". Please use ',' to split datasets and ':' to split parameters of a dataset. 
+We support combination of different datasets and parametrization of calibration datasets by using "--dataset ./tmp.json:concat,NeelNanda/pile-10k:split=train+val:num=256,mbpp:concat=True:num=128". Both local calibration file and huggingface dataset are supported. Please use ',' to split datasets and ':' to split parameters of a dataset.
```
Comment (wenhuach21) [inline examples/language-modeling/README.md]: please detailed concat and split behavior

File: examples/language-modeling/README.md [hunk 4]
```diff
@@ -35,8 +35,8 @@ def customized_data_with_tokenizer(tokenizer, seqlen=2048):
     return tokens
 ~~~
 
-Combination of different datasets has been supported, --dataset "./tmp.json,NeelNanda/pile-10k:train, mbpp:train+validation+test". Please note that samples with sequence length < args.seqlen will be dropped.
-We also support parametrization of calibration datasets by using --dataset "datasetname1:split=train+val:num=256,datasetname2:concat=True:num=128,datasetname3:concat". Please use ',' to split datasets and ':' to split parameters of a dataset. 
+We support combination of different datasets and parametrization of calibration datasets by using "--dataset ./tmp.json:concat,NeelNanda/pile-10k:split=train+val:num=256,mbpp:concat=True:num=128". Both local calibration file and huggingface dataset are supported. Please use ',' to split datasets and ':' to split parameters of a dataset. 
+Please note that samples shorter than args.seqlen will be dropped.
```
Comment (wenhuach21) [inline examples/language-modeling/README.md]: for concat, this is not true
